mdbook-tera-backend/src/tera_renderer/custom_component.rs

```diff
@@ -0,0 +1,133 @@
+use serde::Deserialize;
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::path::{Path, PathBuf};
+
+use tera::Tera;
+
+use crate::{RendererError, Result};
+
+use super::RenderingContext;
+
+fn make_strip_prefix_function() -> impl tera::Function {
+    move |args: &HashMap<String, serde_json::value::Value>| -> tera::Result<tera::Value> {
+        let string = args
+            .get("s")
+            .ok_or_else(|| tera::Error::from(format!("No s argument provided")))?
+            .as_str()
+            .ok_or_else(|| tera::Error::from(format!("S has invalid type, expected string")))?;
+        let prefix = args
+            .get("prefix")
+            .ok_or_else(|| tera::Error::from(format!("No prefix argument provided")))?
+            .as_str()
+            .ok_or_else(|| {
+                tera::Error::from(format!("Prefix has invalid type, expected string"))
+            })?;
+        string
+            .strip_prefix(prefix)
+            .map(|s| tera::Value::String(s.to_owned()))
+            .ok_or_else(|| tera::Error::from(format!("Could not strip prefix")))
+    }
+}
+
+pub struct CustomComponent {
+    template: Tera,
+    name: String,
+    /// Used to generate unique ids for each component to prevent collisions in javascript with query selectors.
+    counter: RefCell<u32>,
+}
+
+impl CustomComponent {
+    pub fn new(name: &str, template_str: &str, dependencies: &[&Self]) -> Result<CustomComponent> {
+        let mut template = Tera::default();
+        for dep in dependencies {
+            template.extend(&dep.template)?;
+        }
+        template.add_raw_template(name, template_str)?;
+        template.register_function("strip_prefix", make_strip_prefix_function());
+
+        Ok(CustomComponent {
+            name: String::from(name),
+            counter: RefCell::new(0),
+            template,
+        })
+    }
+
+    pub fn register_function(&mut self, name: &str, function: impl tera::Function + 'static) {
+        self.template.register_function(name, function);
+    }
+
+    fn create_context(
+        &self,
+        rendering_context: &RenderingContext,
+        attributes: BTreeMap<String, String>,
+    ) -> tera::Context {
+        let counter = self.counter.replace_with(|&mut counter| counter + 1);
+        let mut context = tera::Context::new();
+        context.insert("counter", &counter);
+        context.insert("language", &rendering_context.language);
```

Comment:
This is about i18n, right? :slightly_smiling_face: I'm expecting to see the renderer here dump the entire `book.toml` into the context — a very hands-off approach which allows you to make the renderer completely generic.

Reply:
Yes it's extracted from the book, so we can also remove